mlflow/langchain/api_request_parallel_processor.py

```diff
@@ -75,6 +74,26 @@ class APIRequest:
     request_json: dict
     results: list[tuple[int, str]]
 
+    # https://github.com/langchain-ai/langchain/issues/2222
+    # https://github.com/langchain-ai/langchain/issues/8815
+    # add more formats here if other langchain objects don't serialize
+    def _prepare_to_serialize(self, response: dict):
+        if "intermediate_steps" in response:
+            response["intermediate_steps"] = [
+                {
+                    "tool": agent.tool,
+                    "tool_input": agent.tool_input,
+                    "log": agent.log,
+                    "result": result,
+                }
+                for agent, result in response["intermediate_steps"]
```

Comment:
"intermediate_steps" could be any data structure:
1. https://github.com/langchain-ai/langchain/blob/503c382f882b12eefcd5b73ea46cb964b0f88a90/libs/langchain/langchain/chains/graph_qa/cypher.py#L129
2. https://github.com/langchain-ai/langchain/blob/503c382f882b12eefcd5b73ea46cb964b0f88a90/libs/langchain/langchain/chains/combine_documents/map_rerank.py#L199
3. https://github.com/langchain-ai/langchain/blob/37cb9372c2d65182c228446bf7276fe78e744b95/libs/langchain/langchain/agents/agent.py#L1228

What do you think about the following logic:
1. try-catch using json.dumps to serialize the intermediate_steps.
2. if 1 fails with error, check whether response["intermediate_steps"] is in the format that the above code requires. If so, use the manual serialization logic above. Overwise throws an error saying the intermediate steps are not serializable.

Ideally, we should have a general solution to all the serialization issues. Would you mind filing a JIRA for this?

Reply:
[https://databricks.atlassian.net/browse/ML-34822](https://databricks.atlassian.net/browse/ML-34822)

filed!